### PR TITLE
[onert][gpu_cl] Separate codes as a library which are ported.

### PR DESCRIPTION
**Pull request by Hyunjun85**
- Remove codes which are came from tf-lite's delegates.
- Modify gpu_cl files to use tf-lite delegates.
- Add Cmakelists.txt files to support download and compile library of tf-lite's delegates.
- Add tar.gz which is included delegate functions of Tensorflow v2.4.1.

ref: https://github.com/Samsung/ONE/issues/7460

Signed-off-by:H. Kim hyunjun2.kim@samsung.com

### CODE REVIEW

**Comment by chunseoklee** [comment]
@Hyunjun85 Could you please consider the case in gbs build ? GBS build does not support downloading from web. See tar.gz in ONE/packaging/.

**Comment by Hyunjun85** [comment]
@nnfw-bot test all

**Comment by Hyunjun85** [comment]
@nnfw-bot test x64-builder

**Comment by Hyunjun85** [comment]
@nnfw-bot test x64-builder

**Comment by Hyunjun85** [comment]
@nnfw-bot test android-runtime-builder

**Comment by Hyunjun85** [comment]
@nnfw-bot  test cross-aarch64-runtime-test-release

**Comment by Hyunjun85** [comment]
Rebased with master.

**Comment by Hyunjun85** [comment]
@nnfw-bot test cross-aarch64-runtime-test-release

File: infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+set(TFLITE_GPU_CL_SRCS ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/api.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/buffer.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_command_queue.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/arguments.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_context.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_device.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_event.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_image_format.cc
+                        ${TENSORFLOW_LITE_GPU_SRC_BASE}/cl/cl_kernel.cc
```
Comment (juitem) [inline infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt:16]: 1. There's cmake files for TENSORFLOW_GPU in "infra/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt".
    I think on of them is required.

2.  Do we need keep this cmake on infra/nnfw?

Comment (Hyunjun85) [reply infra/nnfw/cmake/packages/TensorFlowLiteGpu/CMakeLists.txt]: I missed that file, 
actually we don't use that.
So I removed `TENSORFLOW_GPU` in infra/nnfw.

**Comment by Hyunjun85** [comment]
@chunseoklee PTAL.

File: infra/nnfw/cmake/CfgOptionFlags.cmake [hunk 2]
```diff
@@ -54,6 +54,7 @@ option(BUILD_TFLITE_ACCURACY "Build tflite accuracy tool" OFF)
 # Default external libraries source download and build configuration
 #
 option(DOWNLOAD_TENSORFLOW "Download Tensorflow source" ON)
+option(DOWNLOAD_TENSORFLOW_GPU "Download Tensorflow GPU delegate source" ON)
```
Comment (chunseoklee) [inline infra/nnfw/cmake/CfgOptionFlags.cmake:57]: better to be OFF as default to avoid unexpected result(e.g. CI)

File: infra/cmake/packages/Opencl_HeadersConfig.cmake [hunk 3]
```diff
@@ -2,19 +2,24 @@ function(_Opencl_Headers_import)
...
+  endif(TARGET OpenCL_Headers)
 
   if(NOT Opencl_HeadersSource_FOUND)
     message(STATUS "Opencl_Headers: Source not found")
     set(Opencl_Headers_FOUND FALSE PARENT_SCOPE)
     return()
   endif(NOT Opencl_HeadersSource_FOUND)
 
-  add_extdirectory("${Opencl_HeadersSource_DIR}" OPENCL_HEADERS EXCLUDE_FROM_ALL)
+  # We don't need test builds and installs, we only need headers.
```
Comment (chunseoklee) [inline infra/cmake/packages/Opencl_HeadersConfig.cmake:17]: @Hyunjun85 If changes on opencl_header not depends on this PR, please make another PR for this, first.

Comment (Hyunjun85) [reply infra/cmake/packages/Opencl_HeadersConfig.cmake:17]: @chunseoklee We found an issue that I often had to build twice.
And the opencl header has a cycling dependency for that.
So we tried to resolve like above. 

If It is not good for this pr, please let me know once again.

**Comment by Hyunjun85** [comment]
rebased from master and resolved merge conflict.

File: infra/nnfw/cmake/CfgOptionFlags.cmake (hunk 2 above)
Comment (Hyunjun85) [reply infra/nnfw/cmake/CfgOptionFlags.cmake]: I modified.

**Comment by chunseoklee** [comment]
@hseok-oh PTAL though PR is quite huge, But no better way to avoid this size.

**Comment by hseok-oh** [comment]
@Hyunjun85 @chunseoklee This PR looks overkill. When I made issue #7460, I thought that gpc-cl kernel and related codes will be moved into `compute/<new_dir>`, like `cker` to divide compute library. Anyway, I'll check this PR.

**Review by hseok-oh — COMMENTED** [review]
Review is not finished yet, but I found that we can remove some template usage on `gpu-cl` backend (ex. `ClTensorBuilder`). Because of needless template, this PR became more complicate, IMHO. 

And I still have a question which one is better for maintain: remain or remove delegate-related code. (especially for version update or optimize) @Hyunjun85 Do you have any opinion about this?

File: runtime/onert/backend/gpu_cl/Backend.h [hunk 4]
```diff
@@ -27,8 +27,10 @@
 #include "TensorManager.h"
 #include "TensorBuilder.h"
 
-#include "open_cl/Environment.h"
-#include "open_cl/Status.h"
+#include "tensorflow/lite/delegates/gpu/cl/environment.h"
+#include "tensorflow/lite/delegates/gpu/common/status.h"
+
+using namespace tflite::gpu::cl;
```
[the remark this answers is not in the document]
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/Backend.h:33]: I will modify and re-push.

File: runtime/onert/backend/gpu_cl/BackendContext.h [hunk 5]
```diff
@@ -23,7 +23,9 @@
 #include "ConstantInitializer.h"
 #include "KernelGenerator.h"
 #include "TensorBuilder.h"
-#include "open_cl/InferenceContext.h"
+#include "tensorflow/lite/delegates/gpu/cl/inference_context.h"
+
+using namespace tflite::gpu::cl;
```
[the remark this answers is not in the document]
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/BackendContext.h:28]: I will modify and re-push.

File: infra/cmake/packages/TensorFlowGpuSource/patch_for_gpu_cl_build.patch [hunk 6]
```diff
@@ -0,0 +1,292 @@
+diff --git a/tensorflow/lite/delegates/gpu/api.h b/tensorflow/lite/delegates/gpu/api.h
+index 7892d0ce..fae4fb69 100644
+--- a/tensorflow/lite/delegates/gpu/api.h
++++ b/tensorflow/lite/delegates/gpu/api.h
+@@ -43,11 +43,18 @@ limitations under the License.
+ #include "tensorflow/lite/delegates/gpu/common/data_type.h"
+ #include "tensorflow/lite/delegates/gpu/common/status.h"
+ #include "tensorflow/lite/delegates/gpu/common/util.h"
++
++#ifdef TFLITE_GPU_LIB_FIX
```
[the remark this answers is not in the document]
Comment (Hyunjun85) [reply infra/cmake/packages/TensorFlowGpuSource/patch_for_gpu_cl_build.patch:10]: @hseok-oh 
Actually `#ifdef TFLITE_GPU_LIB_FIX` means `if 0`.
Because our final goal is that use original `tflite.so` for gpu delegation.
If we ready, patch_for_gpu_cl_build.patch will be removed also.

So. this PR is a middle step for that.

[the remark this answers is not in the document]
Comment (Hyunjun85) [reply infra/cmake/packages/TensorFlowGpuSource/patch_for_gpu_cl_build.patch:10]: @hseok-oh 
Actually `#ifdef TFLITE_GPU_LIB_FIX` means `if 0`.
Because our final goal is that use original codes of tensorflow-lite for gpu delegation.
If we ready, patch_for_gpu_cl_build.patch will be removed also.

So please look this PR  which it is a middle step for that.

File: runtime/onert/backend/gpu_cl/BackendContext.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/BackendContext.h]: removed the line.

File: runtime/onert/backend/gpu_cl/Backend.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/Backend.h]: removed the line.

**Comment by juitem** [comment]
IMHO,  This pr is a kind of intermediate stage
for maintenance convenience,
1. move codes from gpu_cl/open_cl to /compute/gpu_cl (or somethiing) and close previous issue.
2. improve this pr to include tflite executable functions
   which enables to make 'tflite_gpu_run' and then consider applying improved pr.

**Comment by Hyunjun85** [comment]
@hseok-oh  @juitem 

If we need version update of tensorflow this is much more convenient unless we modify tflite codes. 
And we can easily bring features not supported yet in gpu_cl(gpu_cl still need to port features like quantize, fp16)
and, finally we can also support `tflite-gpu-run` with same lib. (this is a little bit different topic for this pr).

I will check how to remove that patch file for this pr.

**Comment by Hyunjun85** [comment]
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

File: runtime/onert/backend/gpu_cl/ClFunction.h [hunk 7]
```diff
@@ -22,9 +22,11 @@
...
 #include <memory>
 
-#include "open_cl/kernels/GpuOperation.h"
-#include "open_cl/ClCommandQueue.h"
-#include "open_cl/Status.h"
+#include "tensorflow/lite/delegates/gpu/cl/kernels/gpu_operation.h"
+#include "tensorflow/lite/delegates/gpu/cl/cl_command_queue.h"
+#include "tensorflow/lite/delegates/gpu/common/status.h"
+
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/ClFunction.h:29]: Please remove this

File: runtime/onert/backend/gpu_cl/ClMemoryManager.h [hunk 8]
```diff
@@ -21,12 +21,15 @@
...
-#include "open_cl/TensorType.h"
+#include "tensorflow/lite/delegates/gpu/cl/cl_context.h"
+#include "tensorflow/lite/delegates/gpu/common/status.h"
+#include "tensorflow/lite/delegates/gpu/cl/storage_type_util.h"
+#include "tensorflow/lite/delegates/gpu/cl/tensor_type.h"
 #include "util/logging.h"
+#include "ex/InferenceContextEx.h"
+
+using namespace tflite::gpu;
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/ClMemoryManager.h:32]: Same with above

File: runtime/onert/backend/gpu_cl/ClTensorManager.h [hunk 9]
```diff
@@ -19,12 +19,14 @@
...
 
-#include "open_cl/InferenceContext.h"
-#include "open_cl/TensorType.h"
+#include "tensorflow/lite/delegates/gpu/cl/inference_context.h"
+#include "tensorflow/lite/delegates/gpu/cl/tensor_type.h"
 
 #include "ir/OperandInfo.h"
 #include "ir/OperandIndexMap.h"
 
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/ClTensorManager.h:28]: Same

File: runtime/onert/backend/gpu_cl/TensorBuilderHelper.h [hunk 10]
```diff
@@ -15,8 +14,13 @@
...
 
-#ifndef __ONERT_BACKEND_GPU_CL_OPENCL_KERNELS_CONV_COMMON_H__
-#define __ONERT_BACKEND_GPU_CL_OPENCL_KERNELS_CONV_COMMON_H__
+#ifndef __ONERT_BACKEND_GPU_CL_TENSOR_BUILDER_HELPER_H__
+#define __ONERT_BACKEND_GPU_CL_TENSOR_BUILDER_HELPER_H__
+
+#include "absl/status/status.h"
+#include "tensorflow/lite/delegates/gpu/common/shape.h"
+
+using namespace tflite::gpu;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/TensorBuilderHelper.h:23]: Same

File: runtime/onert/backend/gpu_cl/ex/InferenceContextEx.h [hunk 11]
```diff
@@ -15,27 +15,15 @@
...
-#include "GpuObject.h"
-#include "kernels/GpuOperation.h"
-#include "ModelHints.h"
-#include "OpenclWrapper.h"
-#include "Precision.h"
-#include "TensorType.h"
-#include "Model.h"
-#include "InternalTensor.h"
+using namespace tflite::gpu;
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/ex/InferenceContextEx.h:26]: Same

File: runtime/onert/backend/gpu_cl/operand/CLTensor.h [hunk 12]
```diff
@@ -19,9 +19,11 @@
...
 #include "ICLTensor.h"
 
-#include "open_cl/Buffer.h"
-#include "open_cl/ClContext.h"
-#include "open_cl/Tensor.h"
+#include "tensorflow/lite/delegates/gpu/cl/buffer.h"
+#include "tensorflow/lite/delegates/gpu/cl/cl_context.h"
+#include "tensorflow/lite/delegates/gpu/cl/tensor.h"
+
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/operand/CLTensor.h:26]: Same

File: runtime/onert/backend/gpu_cl/operand/ICLTensor.h [hunk 13]
```diff
@@ -19,11 +19,17 @@
...
+#include "tensorflow/lite/delegates/gpu/spi.h"
+#include "tensorflow/lite/delegates/gpu/cl/cl_command_queue.h"
+#include "tensorflow/lite/delegates/gpu/cl/kernels/converter.h"
+#include "tensorflow/lite/delegates/gpu/cl/tensor.h"
+#include "tensorflow/lite/delegates/gpu/cl/environment.h"
+
+#include "TensorBuilderHelper.h"
+
+using namespace tflite::gpu;
+using namespace tflite::gpu::cl;
```
Comment (hseok-oh) [inline runtime/onert/backend/gpu_cl/operand/ICLTensor.h:32]: Same

**Review by hseok-oh — COMMENTED** [review]
Looks good. Please check comment.

File: infra/cmake/packages/TensorFlowGpuSource/patch_for_gpu_cl_build.patch (hunk 6 above)
[the remark this answers is not in the document]
Comment (hseok-oh) [reply infra/cmake/packages/TensorFlowGpuSource/patch_for_gpu_cl_build.patch:10]: I see.

File: runtime/onert/backend/gpu_cl/ClFunction.h (hunk 7 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/ClFunction.h:29]: Removed.

File: runtime/onert/backend/gpu_cl/ClMemoryManager.h (hunk 8 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/ClMemoryManager.h:32]: Removed.

File: runtime/onert/backend/gpu_cl/ClTensorManager.h (hunk 9 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/ClTensorManager.h:28]: Removed.

File: runtime/onert/backend/gpu_cl/TensorBuilderHelper.h (hunk 10 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/TensorBuilderHelper.h:23]: Removed.

File: runtime/onert/backend/gpu_cl/operand/CLTensor.h (hunk 12 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/operand/CLTensor.h:26]: Removed.

File: runtime/onert/backend/gpu_cl/operand/ICLTensor.h (hunk 13 above)
Comment (Hyunjun85) [reply runtime/onert/backend/gpu_cl/operand/ICLTensor.h:32]: Removed.

**Comment by Hyunjun85** [comment]
> Looks good. Please check comment.

I applied your comments.

**Review by hseok-oh — APPROVED** [review]

**Comment by Hyunjun85** [comment]
@chunseoklee PTAL.

**Review by juitem — APPROVED** [review]
LGTM, don't forget remove patch file on next commit.

**Review by chunseoklee — APPROVED** [review]
